beacon_node/beacon_chain/src/shuffling_cache.rs

```diff
@@ -127,13 +145,60 @@ impl ShufflingCache {
             // worth two in the promise-bush!
             .map_or(true, CacheItem::is_promise)
         {
-            self.cache.put(
+            self.insert_cache_item(
                 key,
                 CacheItem::Committee(committee_cache.to_arc_committee_cache()),
             );
         }
     }
 
+    /// Prunes the cache first before inserting a new cache item.
+    fn insert_cache_item(&mut self, key: AttestationShufflingId, cache_item: CacheItem) {
+        self.prune_cache();
+        self.cache.insert(key, cache_item);
+    }
+
+    /// Prunes the `cache` to keep the size below the `cache_size` limit, based on the following preferences:
+    /// - Entries from more recent epochs are preferred over older ones.
+    /// - "Enshrined" shuffling (i.e. shuffling id for the head block at the current wall-clock epoch)
+    ///   must not be pruned.
+    fn prune_cache(&mut self) {
+        if self.cache.len() >= self.cache_size {
+            let prune_count = self.cache.len() - self.cache_size + 1;
+            let shuffling_ids_to_prune = self
+                .cache
+                .keys()
+                .sorted_by_key(|key| key.shuffling_epoch)
```

Comment:
Sorting the hashmap by epoch and breaking ties by random hashmap ordering seems good to me.

I don't think we really have a good justification for preferring shufflings derived from ancestors of the head vs shufflings that conflict with the head. Although the former are probably more likely in practice, the latter are more useful in the case of multiple long forks (hopefully unlikely).